### PR TITLE
change http code to 400 for invalid resourceVersion for GET rest apis

### LINKED ISSUE #101350: Querying several resources via API with invalid `resourceVersion` query parameter causes internal server error
#### What happened:
Hello @viralpoetry and I were fuzzing the k8s (via [openapi-fuzzer](https://github.com/matusf/openapi-fuzzer)) and found out following bug. Querying several resources via API with invalid `resourceVersion` query parameter causes internal server error. Affected resources: 
- `api-v1-namespaces-{namespace}-configmaps/`
- `api-v1-namespaces-{namespace}-endpoints/`
- `api-v1-namespaces-{namespace}-events/`
- `api-v1-namespaces-{namespace}-limitranges/`
- `api-v1-namespaces-{namespace}-persistentvolumeclaims/`
- `api-v1-namespaces-{namespace}-pods/`
- `api-v1-namespaces-{namespace}-podtemplates/`
- `api-v1-namespaces-{namespace}-replicationcontrollers/`
- `api-v1-namespaces-{namespace}-resourcequotas/`
- `api-v1-namespaces-{namespace}-secrets/`
- `api-v1-namespaces-{namespace}-serviceaccounts/`
- `api-v1-namespaces-{namespace}-services/`

#### What you expected to happen:
Response with non-500 HTTP status code

#### How to reproduce it (as minimally and precisely as possible):
We were fuzzing k8s locally via minikube. 
1. start minukube
```
minikube start --driver virtualbox
```
2. install ca certificates
```
sudo cp ~/.minikube/ca.pem /usr/local/share/ca-certificates/
sudo update-ca-certificates
```
3. get k8s IP address
```
kubectl config view --minify -o jsonpath='{.clusters[0].cluster.server}'
```
4. add the IP to `/etc/hosts` as `minikubeca`
5. get access token and export it as `token`
```
kubectl create clusterrolebinding permissive-binding \
  --clusterrole=cluster-admin \
  --user=admin \
  --user=kubelet \
  --group=system:serviceaccounts

kubectl create serviceaccount openapifuzz
kubectl get serviceaccounts openapifuzz -o json | jq -r .secrets[].name
kubectl get secret openapifuzz-token-sn9qm -o json | jq -r .data.token
 export token=<retrieved-token>
```
6. send requests
```
# configmaps
curl -X GET 'https://minikubeca:8443/api/v1/namespaces/x/configmaps?resourceVersion=x' -H "Authorization: Bearer $token"

{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "resourceVersion: Invalid value: \"x\": strconv.ParseUint: parsing \"x\": invalid syntax",
  "code": 500
}

# endpoints
curl -X GET -H "authorization: bearer $token" 'https://minikubeca:8443/api/v1/namespaces/x/endpoints?resourceVersion=x'
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "resourceVersion: Invalid value: \"x\": strconv.ParseUint: parsing \"x\": invalid syntax",
  "code": 500
}

# events
curl -X GET -H "authorization: bearer $token" 'https://minikubeca:8443/api/v1/namespaces/x?events&resourceVersion=x'
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "resourceVersion: Invalid value: \"x\": strconv.ParseUint: parsing \"x\": invalid syntax",
  "code": 500
}

# limit ranges
curl -X GET -H "authorization: bearer $token" 'https://minikubeca:8443/api/v1/namespaces/x/limitranges?resourceVersion=x'
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "resourceVersion: Invalid value: \"x\": strconv.ParseUint: parsing \"x\": invalid syntax",
  "code": 500
}

# persistent volume claims
curl -X GET -H "authorization: bearer $token" 'https://minikubeca:8443/api/v1/namespaces/x/persistentvolumeclaims?resourceVersion=x'
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "resourceVersion: Invalid value: \"x\": strconv.ParseUint: parsing \"x\": invalid syntax",
  "code": 500
}

# pods
curl -X GET -H "authorization: bearer $token" 'https://minikubeca:8443/api/v1/namespaces/x/pods?resourceVersion=x'
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "resourceVersion: Invalid value: \"x\": strconv.ParseUint: parsing \"x\": invalid syntax",
  "code": 500
}

# pod templates
curl -X GET -H "authorization: bearer $token" 'https://minikubeca:8443/api/v1/namespaces/x/podtemplates?resourceVersion=x'
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "resourceVersion: Invalid value: \"x\": strconv.ParseUint: parsing \"x\": invalid syntax",
  "code": 500
}

# replication controllers
curl -X GET -H "authorization: bearer $token" 'https://minikubeca:8443/api/v1/namespaces/x/replicationcontrollers?resourceVersion=x'
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "resourceVersion: Invalid value: \"x\": strconv.ParseUint: parsing \"x\": invalid syntax",
  "code": 500
}

# resource quotas
curl -X GET -H "authorization: bearer $token" 'https://minikubeca:8443/api/v1/namespaces/x/resourcequotas?resourceVersion=x'
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "resourceVersion: Invalid value: \"x\": strconv.ParseUint: parsing \"x\": invalid syntax",
  "code": 500
}

# secrets
curl -X GET -H "authorization: bearer $token" 'https://minikubeca:8443/api/v1/namespaces/x/secrets?resourceVersion=x'
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "resourceVersion: Invalid value: \"x\": strconv.ParseUint: parsing \"x\": invalid syntax",
  "code": 500
}

# service accounts
curl -X GET -H "authorization: bearer $token" 'https://minikubeca:8443/api/v1/namespaces/x/serviceaccounts?resourceVersion=x'
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "resourceVersion: Invalid value: \"x\": strconv.ParseUint: parsing \"x\": invalid syntax",
  "code": 500
}

# services
curl -X GET -H "authorization: bearer $token" 'https://minikubeca:8443/api/v1/namespaces/x/services?resourceVersion=x'
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {
    
  },
  "status": "Failure",
  "message": "resourceVersion: Invalid value: \"x\": strconv.ParseUint: parsing \"x\": invalid syntax",
  "code": 500
}
```

#### Environment:
- Kubernetes version (use `kubectl version`): 
```
clientVersion:
  buildDate: "2021-04-08T16:31:21Z"
  compiler: gc
  gitCommit: cb303e613a121a29364f75cc67d3d580833a7479
  gitTreeState: clean
  gitVersion: v1.21.0
  goVersion: go1.16.1
  major: "1"
  minor: "21"
  platform: linux/amd64
serverVersion:
  buildDate: "2021-01-13T13:20:00Z"
  compiler: gc
  gitCommit: faecb196815e248d3ecfb03c680a4507229c2a56
  gitTreeState: clean
  gitVersion: v1.20.2
  goVersion: go1.15.5
  major: "1"
  minor: "20"
  platform: linux/amd64
```
- Cloud provider or hardware configuration: minikube with virtual-box driver
- OS (e.g: `cat /etc/os-release`): Ubuntu 18.04.5 LTS (bionic)
- Kernel (e.g. `uname -a`): Linux pine 4.15.0-140-generic #144-Ubuntu SMP Fri Mar 19 14:12:35 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: downloaded precompiled binary
- minikube version: v1.19.0 (commit: 15cede53bdc5fe242228853e737333b09d4336b5)

### PR DESCRIPTION
**Pull request by pacoxu**
#### What type of PR is this?
/kind bug

#### Which issue(s) this PR fixes:
Fixes #101350

#### Special notes for your reviewer:
NewInvalidError is 500?
change to "errors.NewBadRequest" to be 400

#### Does this PR introduce a user-facing change?
```release-note
return 400 for invalid resourceVersion in GET REST APIs
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go [hunk 1]
```diff
@@ -99,6 +100,14 @@ func GetResource(r rest.Getter, scope *RequestScope) http.HandlerFunc {
 						return nil, errors.NewBadRequest("the export parameter, deprecated since v1.14, is no longer supported")
 					}
 				}
+				// resourceVersion should be "" or int
+				if len(values["resourceVersion"]) > 0 {
+					_, err := strconv.ParseUint(values["resourceVersion"][0], 10, 64)
```
Comment (yliaog) [inline staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go:105]: resourceVersion should be an opaque string, it happened to be 'int' with the current use of etcd as datastore.

Comment (pacoxu) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go:105]: I just checked the type definition. You are right. 
So does this mean we should not fix #101350? @yliaog 

https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions
> Resource versions must be treated as opaque by clients and passed unmodified back to the server.

https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go#L47-L50

Comment (pacoxu) [reply staging/src/k8s.io/apiserver/pkg/endpoints/handlers/get.go]: Only one datastore type is supported now:  etcd3.

>       --storage-backend string                   The storage backend for persistence. Options: 'etcd3' (default).

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/api_object_versioner.go [hunk 2]
```diff
@@ -94,11 +95,11 @@ func (a APIObjectVersioner) ParseResourceVersion(resourceVersion string) (uint64
 	}
 	version, err := strconv.ParseUint(resourceVersion, 10, 64)
 	if err != nil {
-		return 0, storage.NewInvalidError(field.ErrorList{
+		return 0, errors.NewBadRequest(
```
Comment (yliaog) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/api_object_versioner.go:98]: this looks good to me, @deads2k is there any historical reason NewBadRequest was not used initially before?

Comment (pacoxu) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/api_object_versioner.go:98]: As this method is added by Daniel and approved by Jordon,
/cc liggitt lavalamp

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/api_object_versioner.go:98]: this utility method returning an API error is a little weird... normally, the thing calling this method in an API context would wrap in an API error... it looks like we do that in ~1/3 of the places we call this

**Comment by lavalamp** [comment]
Maybe it's too late, but personally I would prefer not to give the client hints about why the RV is invalid, I don't want them to randomly search integers, and I want to leave the door open for different RV formats in the future.

I agree with the 400 status code though.

**Pull request closed without merging**